examples/computer_vision/vision_transformer/det_callback.py

```diff
@@ -0,0 +1,187 @@
+        import determined
+        self._det = determined
+        distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
+
+        self.filter_metrics = filter_metrics
+        self.user_checkpoint_metadata = checkpoint_metadata
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_metrics = {}
+        self.searcher_metric = self._det.get_cluster_info().trial._config['searcher']['metric']
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+    def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, model=None, logs=None,
+               **kwargs):
+        if state.is_world_process_zero:
+            metrics, metric_type = self.get_metrics(logs)
+            if metric_type == TRAIN:
+                self.core_context.train.report_training_metrics(steps_completed=state.global_step, metrics=metrics)
+            elif metric_type == EVAL:
+                self.core_context.train.report_validation_metrics(steps_completed=state.global_step, metrics=metrics)
+            else:
+                logging.warning(f"Metrics not reported: metric type = {metric_type}.")
+
+            self.last_metrics.update(metrics)
+
+    def get_metrics(self, logs: typing.Dict) -> typing.Tuple[typing.Dict, str]:
+        metrics = logs
+        metric_type = get_metric_type(logs)
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics, metric_type
+
+    def on_save(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        info = self._det.get_cluster_info()
+        if info is None:
+            # TODO: modify to support local mode
+            logging.warning('ClusterInfo is None: not running in a task. Skip saving.')
+            return
+
+        if state.is_world_process_zero:
+            save_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+            det_checkpoint_metadata = {"steps_completed": state.global_step, "trial_id": info.trial.trial_id}
+
+            if self.tokenizer_options is not None:
+                det_checkpoint_metadata['tokenizer_options'] = self.tokenizer_options
+
+            if self.tokenizer is not None:
+                self.tokenizer.save_pretrained(os.path.join(save_path, "tokenizer"))
+
+            if self.user_checkpoint_metadata is not None:
+                self._on_save_user_data(save_path)
+
+            self.core_context.checkpoint.upload(save_path, det_checkpoint_metadata)
```

Comment:
this could fail if a user is using the deepspeed backend with sharded checkpoint that need to be take across multiple nodes in which case only files on the rank 0 node would get uploaded.

Reply:
Added uploading checkpoints from multiple nodes.